Add prefix_length option for misspellings

toyray opened the pull request:
Added prefix_length option for misspellings to disable fuzzy matching for short words (keeping prefix_length for lack of a better name).

Main gotcha is when a query term has the same length as prefix_length whereby it becomes an exact match query, hence the extra test case.

This should fix #410, but defaulting prefix_length to 0 if unspecified for backwards compatibility.

Let me know if you want me to submit  PR to update README for this.

ankane commented:
Thanks Roy, this is useful!  The case when prefix_length is the same length as the term seems counter-intuitive (seems like `ap` with prefix_length of 2 should match `api`), but I'm not sure there's an easy way around it.  An example in the readme would great, and it'd be good to mention this caveat.

toyray commented:
I'll see if there's a way to tweak the prefix_length based on the search term to make the behavior more intuitive.

Anyways, I'll add a commit to document this in the README and rebase this for easier merging later.

toyray commented:
I added a way to handle the case where prefix length = term length by using the term length - 1 so the misspellings still work for those cases. This should also work for the "or" operator since the length of the shortest token is always used.

Updated the test cases accordingly (let me know if there are other cases that should be covered)

I'm not too sure if below is the best way of getting the shortest token from the query term.

```ruby
search2_prefix_length = [
    (searchkick_index.tokens(term, analyzer: "searchkick_search2")
        .min_by(&:length)
        .try(:length) || 1
    ) - 1,
     prefix_length
].min
```

the other alternative I had was

``` ruby
search2_prefix_length = [
    term.split(" ").min_by(&:length).try(:length) || 1) - 1,
     prefix_length
].min
```

Let me know what you think and if this is good, I'll add another commit to update the README

ankane commented:
With the `prefix_length - 1` approach, `ak` will match `ap` with a prefix length of 2, which is also counterintuitive.  I'd say we go with the original approach.

toyray commented:
Yup, I'll rebase without the `prefix_length - 1` commit and add a commit to update the README.

Thanks for the feedback, Andrew.

ankane commented:
Merged, thanks @toyray!
